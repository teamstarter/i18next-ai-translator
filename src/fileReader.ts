import fs from 'fs';
import path from 'path';
import debug from 'debug';
import { TranslationFile, UntranslatedKey } from './types';

const log = debug('i18next-ai-translator:fileReader');

export function readTranslationFiles(localesFolderPath: string): TranslationFile[] {
  log('Reading translation files from: %s', localesFolderPath);

  if (!fs.existsSync(localesFolderPath)) {
    log('Error: Folder does not exist: %s', localesFolderPath);
    throw new Error(`The folder ${localesFolderPath} does not exist`);
  }

  const files = fs.readdirSync(localesFolderPath);
  log('Found %d files in folder', files.length);

  const jsonFiles = files.filter(file => file.endsWith('.json'));
  log('Filtered to %d JSON files', jsonFiles.length);

  const translationFiles: TranslationFile[] = [];

  for (const file of jsonFiles) {
    const filePath = path.join(localesFolderPath, file);
    const locale = path.basename(file, '.json');

    log('Reading file: %s (locale: %s)', file, locale);

    const content = fs.readFileSync(filePath, 'utf-8');
    const translations = JSON.parse(content);

    translationFiles.push({
      locale,
      filePath,
      translations,
    });

    log('Successfully loaded %s with %d root keys', file, Object.keys(translations).length);
  }

  log('Total translation files loaded: %d', translationFiles.length);
  return translationFiles;
}

function findKeysRecursive(
  obj: Record<string, unknown>,
  currentPath: string[] = [],
  untranslatedValue: string
): string[][] {
  const untranslatedPaths: string[][] = [];

  for (const [key, value] of Object.entries(obj)) {
    const newPath = [...currentPath, key];

    if (typeof value === 'object' && value !== null) {
      const nestedPaths = findKeysRecursive(
        value as Record<string, unknown>,
        newPath,
        untranslatedValue
      );
      untranslatedPaths.push(...nestedPaths);
    } else if (value === untranslatedValue) {
      log('Found untranslated key: %s', newPath.join('.'));
      untranslatedPaths.push(newPath);
    }
  }

  return untranslatedPaths;
}

export function findUntranslatedKeys(
  translationFiles: TranslationFile[],
  untranslatedValue: string = '__STRING_NOT_TRANSLATED__'
): UntranslatedKey[] {
  log('Searching for untranslated keys with value: "%s"', untranslatedValue);
  log('Processing %d translation files', translationFiles.length);

  const untranslatedKeys: UntranslatedKey[] = [];

  for (const file of translationFiles) {
    log('Analyzing file: %s (locale: %s)', file.filePath, file.locale);

    const keyPaths = findKeysRecursive(file.translations, [], untranslatedValue);
    log('Found %d untranslated keys in %s', keyPaths.length, file.locale);

    for (const keyPath of keyPaths) {
      untranslatedKeys.push({
        keyPath: keyPath,
        locale: file.locale,
        filePath: file.filePath,
      });
    }
  }

  log('Total untranslated keys found: %d', untranslatedKeys.length);
  return untranslatedKeys;
}
