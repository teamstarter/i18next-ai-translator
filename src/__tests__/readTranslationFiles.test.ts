import fs from 'fs';
import path from 'path';
import { readTranslationFiles } from '../fileReader';

const TEST_DIR = path.join(__dirname, '__test-locales-read__');

describe('readTranslationFiles', () => {
  it('should read a single JSON translation file', () => {
    fs.mkdirSync(TEST_DIR, { recursive: true });
    const enContent = { welcome: 'Welcome', home: 'Home' };
    fs.writeFileSync(path.join(TEST_DIR, 'en.json'), JSON.stringify(enContent));

    const result = readTranslationFiles(TEST_DIR);

    expect(result).toHaveLength(1);
    expect(result[0].locale).toBe('en');
    expect(result[0].translations).toEqual(enContent);

    fs.rmSync(TEST_DIR, { recursive: true, force: true });
  });

  it('should read multiple JSON translation files', () => {
    fs.mkdirSync(TEST_DIR, { recursive: true });
    fs.writeFileSync(path.join(TEST_DIR, 'en.json'), JSON.stringify({ key: 'English' }));
    fs.writeFileSync(path.join(TEST_DIR, 'fr.json'), JSON.stringify({ key: 'Français' }));

    const result = readTranslationFiles(TEST_DIR);

    expect(result).toHaveLength(2);
    expect(result.map(f => f.locale)).toContain('en');
    expect(result.map(f => f.locale)).toContain('fr');

    fs.rmSync(TEST_DIR, { recursive: true, force: true });
  });

  it('should ignore non-JSON files', () => {
    fs.mkdirSync(TEST_DIR, { recursive: true });
    fs.writeFileSync(path.join(TEST_DIR, 'en.json'), JSON.stringify({ key: 'value' }));
    fs.writeFileSync(path.join(TEST_DIR, 'readme.txt'), 'Some text');

    const result = readTranslationFiles(TEST_DIR);

    expect(result).toHaveLength(1);
    expect(result[0].locale).toBe('en');

    fs.rmSync(TEST_DIR, { recursive: true, force: true });
  });

  it('should return empty array when folder is empty', () => {
    fs.mkdirSync(TEST_DIR, { recursive: true });

    const result = readTranslationFiles(TEST_DIR);

    expect(result).toEqual([]);

    fs.rmSync(TEST_DIR, { recursive: true, force: true });
  });

  it('should throw error when folder does not exist', () => {
    const nonExistentPath = path.join(TEST_DIR, 'does-not-exist');

    expect(() => readTranslationFiles(nonExistentPath)).toThrow(
      `The folder ${nonExistentPath} does not exist`
    );
  });

  it('should include correct file path in result', () => {
    // Préparer
    fs.mkdirSync(TEST_DIR, { recursive: true });
    fs.writeFileSync(path.join(TEST_DIR, 'en.json'), JSON.stringify({ key: 'value' }));

    const result = readTranslationFiles(TEST_DIR);

    expect(result[0].filePath).toBe(path.join(TEST_DIR, 'en.json'));

    fs.rmSync(TEST_DIR, { recursive: true, force: true });
  });

  it('should match snapshot for translation files structure', () => {
    fs.mkdirSync(TEST_DIR, { recursive: true });
    const content = {
      welcome: 'Welcome',
      nested: {
        key: 'value',
      },
    };
    fs.writeFileSync(path.join(TEST_DIR, 'en.json'), JSON.stringify(content));

    const result = readTranslationFiles(TEST_DIR);

    expect(result).toHaveLength(1);
    expect(result[0].filePath).toBe(path.join(TEST_DIR, 'en.json'));

    expect({
      locale: result[0].locale,
      translations: result[0].translations,
    }).toMatchSnapshot();

    fs.rmSync(TEST_DIR, { recursive: true, force: true });
  });
});
