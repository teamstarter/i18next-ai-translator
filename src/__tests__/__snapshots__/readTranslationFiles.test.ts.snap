// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`readTranslationFiles should match snapshot for translation files structure 1`] = `
[
  {
    "filePath": "/Users/Bleza/Documents/i18next-ai-translator/src/__tests__/__test-locales-read__/en.json",
    "locale": "en",
    "translations": {
      "nested": {
        "key": "value",
      },
      "welcome": "Welcome",
    },
  },
]
`;
